File: ios-app/BayitPlusApp/ViewModels/WidgetDockViewModel.swift
import BayitCore
import Foundation
import Observation

/// Lightweight view model for the global floating widget dock.
/// Loads the user's active widgets and manages dock visibility state.
/// Tracks which widgets are minimized (shown in dock) vs restored (shown as floating PiP).
/// Used by MainTabView to render PiPWidgetManagerView across all tabs.
@MainActor
@Observable
final class WidgetDockViewModel {
    private(set) var widgets: [WidgetItem] = []
    private(set) var isLoading = false
    var isDockVisible = true

    /// IDs of widgets that have been restored from the dock (not minimized).
    private(set) var restoredWidgetIds: Set<String> = []

    private let repository: any WidgetRepository
    private let logger = BayitLogger(category: "WidgetDock")
    private var persistTask: Task<Void, Never>?

    init(repository: any WidgetRepository, initiallyVisible: Bool = false) {
        self.repository = repository
        isDockVisible = initiallyVisible
    }

    /// Show the dock explicitly.
    @MainActor
    func showDock() {
        isDockVisible = true
    }

    /// Widgets currently minimized in the dock (excludes restored ones).
    var minimizedWidgets: [WidgetItem] {
        widgets.filter { !restoredWidgetIds.contains($0.id) }
    }

    /// Widgets currently restored as floating PiP windows.
    var restoredWidgets: [WidgetItem] {
        widgets.filter { restoredWidgetIds.contains($0.id) }
    }

    /// Load the user's active widgets for the dock.
    @MainActor
    func loadWidgets() async {
        guard !isLoading else { return }
        isLoading = true

        do {
            let response = try await repository.fetchMyWidgets()
            widgets = response.items.filter { $0.isVisible != false }

            // Don't auto-restore widgets into the floating sidebar on cold launch.
            // The sidebar should only appear when the user explicitly restores a
            // widget from the dock during the current session.
            restoredWidgetIds = []

            // Only show dock if there are widgets sitting in minimized state
            if !isDockVisible {
                isDockVisible = !minimizedWidgets.isEmpty
            }

            logger.info("Loaded \(widgets.count) dock widgets")
        } catch {
            logger.error("Failed to load dock widgets: \(error.localizedDescription)")
        }

        isLoading = false
    }

    /// Toggle a widget between minimized (in dock) and restored (floating PiP).
    @MainActor
    func toggleMinimize(widgetId: String) {
        let isCurrentlyRestored = restoredWidgetIds.contains(widgetId)
        let newIsMinimized = isCurrentlyRestored

        // Optimistic update
        if isCurrentlyRestored {
            restoredWidgetIds.remove(widgetId)
        } else {
            restoredWidgetIds.insert(widgetId)
        }

        // Persist to backend
        persistTask = Task {
            do {
                _ = try await repository.toggleMinimize(
                    widgetId: widgetId,
                    isMinimized: newIsMinimized
                )
            } catch {
                // Revert on failure
                await MainActor.run {
                    if newIsMinimized {
                        restoredWidgetIds.insert(widgetId)
                    } else {
                        restoredWidgetIds.remove(widgetId)
                    }
                }
                logger.error("Failed to toggle minimize: \(error.localizedDescription)")
            }
        }
    }

    /// Minimize a restored widget back into the dock.
    @MainActor
    func minimizeWidget(widgetId: String) {
        guard restoredWidgetIds.contains(widgetId) else { return }
        toggleMinimize(widgetId: widgetId)
    }

    /// Close a widget from the sidebar, deactivating it (no longer shown in UI).
    @MainActor
    func closeWidget(widgetId: String) {
        minimizeWidget(widgetId: widgetId)
    }

    /// Remove a widget from the sidebar display (tvOS: dismisses without persisting minimize state).
    @MainActor
    func dismissFromSidebar(widgetId: String) {
        widgets.removeAll { $0.id == widgetId }
    }

    /// Toggle dock visibility.
    @MainActor
    func toggleDock() {
        isDockVisible.toggle()
    }

    /// Hide the dock.
    @MainActor
    func hideDock() {
        isDockVisible = false
    }
}

File: android-app/app/src/main/java/tv/bayit/plus/ui/viewmodel/WidgetDockViewModel.kt
package tv.bayit.plus.ui.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import tv.bayit.plus.core.common.BayitResult
import tv.bayit.plus.core.data.repository.WidgetRepository
import tv.bayit.plus.core.model.WidgetItem
import javax.inject.Inject

/**
 * ViewModel managing the global floating widget dock state.
 * Loads user's active widgets and tracks minimize/restore state.
 * Matches iOS WidgetDockViewModel functionality.
 */
@HiltViewModel
class WidgetDockViewModel @Inject constructor(
    private val widgetRepository: WidgetRepository,
) : ViewModel() {

    private val _uiState = MutableStateFlow(WidgetDockUiState())
    val uiState: StateFlow<WidgetDockUiState> = _uiState.asStateFlow()

    private val restoredWidgetIds = mutableSetOf<String>()

    init {
        loadWidgets()
    }

    fun loadWidgets() {
        if (_uiState.value.isLoading) return

        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true)

            when (val result = widgetRepository.getActiveWidgets()) {
                is BayitResult.Success -> {
                    val widgets = result.data.filterIsInstance<WidgetItem>()
                        .filter { it.isVisible != false }

                    restoredWidgetIds.clear()
                    restoredWidgetIds.addAll(
                        widgets.filter { it.isMinimized == false }.map { it.id }
                    )

                    _uiState.value = _uiState.value.copy(
                        widgets = widgets,
                        isLoading = false,
                        error = null
                    )
                }

                is BayitResult.Error -> {
                    _uiState.value = _uiState.value.copy(
                        isLoading = false,
                        error = result.message
                    )
                }

                is BayitResult.Loading -> {
                    // Already loading, no-op
                }
            }
        }
    }

    fun toggleMinimize(widgetId: String) {
        val isCurrentlyRestored = restoredWidgetIds.contains(widgetId)

        if (isCurrentlyRestored) {
            restoredWidgetIds.remove(widgetId)
        } else {
            restoredWidgetIds.add(widgetId)
        }

        _uiState.value = _uiState.value.copy(
            restoredWidgetIds = restoredWidgetIds.toSet()
        )

        viewModelScope.launch {
            val newIsMinimized = isCurrentlyRestored
            val result = widgetRepository.toggleMinimize(widgetId, newIsMinimized)

            if (result is BayitResult.Error) {
                if (isCurrentlyRestored) {
                    restoredWidgetIds.add(widgetId)
                } else {
                    restoredWidgetIds.remove(widgetId)
                }
                _uiState.value = _uiState.value.copy(
                    restoredWidgetIds = restoredWidgetIds.toSet()
                )
            }
        }
    }

    fun minimizeWidget(widgetId: String) {
        if (restoredWidgetIds.contains(widgetId)) {
            toggleMinimize(widgetId)
        }
    }

    fun toggleDock() {
        _uiState.value = _uiState.value.copy(
            isDockVisible = !_uiState.value.isDockVisible
        )
    }

    fun hideDock() {
        _uiState.value = _uiState.value.copy(isDockVisible = false)
    }

    fun getMinimizedWidgets(): List<WidgetItem> {
        return _uiState.value.widgets.filter { widget ->
            !restoredWidgetIds.contains(widget.id)
        }
    }

    fun getRestoredWidgets(): List<WidgetItem> {
        return _uiState.value.widgets.filter { widget ->
            restoredWidgetIds.contains(widget.id)
        }
    }
}

data class WidgetDockUiState(
    val widgets: List<WidgetItem> = emptyList(),
    val restoredWidgetIds: Set<String> = emptySet(),
    val isDockVisible: Boolean = true,
    val isLoading: Boolean = false,
    val error: String? = null,
)
